GoogleForCreators/web-stories-wp
# Design System: Search Added to Storybook

## Summary
Search component based on [Figma ](https://www.figma.com/file/bMhG3KyrJF8vIAODgmbeqT/Design-System?node-id=1533%3A260040). To be used as generic search component (dashboard for stories, editor for images) 

![keyboard search](https://user-images.githubusercontent.com/10720454/106834097-014b5e00-6652-11eb-8c8d-8cbc299cae04.gif)

## Relevant Technical Choices
- Reuses `components/menu` to handle menu functionality. Updates `Menu` to have a forwardedRef. We need this in the search to pass focus back to the input to avoid focus traps where you can't go back to the input to narrow your options via keyboard. 
- `Search` is separated from `DropDown` because the way the menus are accessed and values are updated between the two are pretty different. Menu's shared. 
- I had tried to combine the generic search (where you can type anything into the input and it'll stay regardless of if it's a valid option presented in the menu) and the font search in the editor (which uses the input as a guidepost for limiting options but is itself not a valid option) and that made everything a little too convoluted. 
- `useSearch` is a bit like `useDropDown` - it's where all logic for state of search is stored - all state and useEffects.
- I separated `selectedValue` (used in dropDown) from the inputValue so that selectedValue doesn't update when input changes. There's a new `handleSearchValueChange` callback that gives parent access to inputValue. I did this so that how search and filtering of options is handled isn't up to the typeahead but the parent so that we can reuse the UI and untangle logic from visual. 
- We've updated how keyboard focus styles are handled to decrease reliance on javascript and utilize `:focus-visible` - the input here is the first case where this gets a little weird, on focus of the input we're updating some state attached to the search that let's the component know that a menu could be rendered and to adjust accordingly. This makes the input rerender, losing the class associated with that keyboard style. I got a little clever to give the input the styles it needed by referencing `::placeholder` and using `&[aria-expanded='true']` to avoid needing the `isOpen` prop. 

## To-do

- [x] Props clean up 
- [x] proof of concept - dashboard
- [x] proof of concept - editor 
- [x] Fix tests (this branch has had a long life)


## User-facing changes

- None, exists in storybook 

## Testing Instructions

(Run storybook locally and go to design-system/components/search)
- See dark and light theme variations with lots of knobs to change props. 
- Try changing value, picking new options from the menu, changing knobs.

Functionality: 
- Focusing on input should shift the search icon to the left. If the input's empty no icon should be visible to the right hand side.
- As soon as the input has a value length > 0 available menu options will be allowed to render. If there are none then the empty results message will show where the menu would be. 
- If there is value in input and menu is open the 'x' clear button should be present
- if there is a value in input and menu is closed there is a chevron present. 
- search icon is always present
- When input is inactive and has no value, placeholder and search icon are centered.
- You should be able to enter the expanded menu by arrowDown from input and you should be able to leave the menu and return to the input with arrowUp. 
- If you have been typing text into the input and hit tab, you should land on the clear button. 
- Clicking on a clear button should clear the input and thereby close the menu.


---



Addresses #4951
